src/evolution.jl

```diff
@@ -67,124 +131,68 @@ function evolve(s, τ, N, B, L, N_sites, Δx, Δm², p, x, Δp, theta_nu, ψ, sh
         # It truncates the MPS according to the set cutoff and maxdim for all the non-nearest-neighbor gates.
         ψ = apply(gates, ψ; cutoff, maxdim)
-        # ψ = tdvp(H, ψ,  -im *τ;   nsweeps=1,
-        # reverse_step=true,outputlevel=1)
 
         # The normalize! function is used to ensure that the MPS is properly normalized after each application of the time evolution gates. 
         # This is necessary to ensure that the MPS represents a valid quantum state.
         normalize!(ψ)
-    end
-    t_array = 0.0:τ:ttotal
-
-    # After the time evolution loop, calculate and print the growth rate from the ratio if both values have been captured
-    if ρₑμ_at_t1 !== nothing && ρₑμ_at_t2 !== nothing
-        global Im_Ω = (1/Δt ) * log(ρₑμ_at_t2 / ρₑμ_at_t1)
-        println("Growth rate of flavor coherence of ρₑμ at t2 to ρₑμ at t1: $Im_Ω")
-    else
-        println("ρₑμ was not captured at both t1 and t2.")
-    end
 
-    if save_data
-        save_data = isdir(datadir) || mkpath(datadir)
-        # Writing data to files with corresponding headers
-        fname1 = joinpath(datadir, "t_<Sz>_<Sy>_<Sx>.dat")
-        writedlm(fname1, [t_array Sz_array Sy_array Sx_array])
-        
-        fname2 = joinpath(datadir, "t_probsurv.dat")
-        writedlm(fname2, [t_array prob_surv_array])
-        
-        fname3 = joinpath(datadir, "t_xsiteval.dat")
-        writedlm(fname3, [t_array x_values])
-        
-        fname4 = joinpath(datadir, "t_pxsiteval.dat")
-        writedlm(fname4, [t_array pₓ_values])
-        
-        fname5 = joinpath(datadir, "t_ρₑₑ.dat")
-        writedlm(fname5, [t_array ρₑₑ_array])
-        
-        fname6 = joinpath(datadir, "t_ρ_μμ.dat")
-        writedlm(fname6, [t_array ρ_μμ_array])
-        
-        fname7 = joinpath(datadir, "t_ρₑμ.dat")
-        writedlm(fname7, [t_array ρₑμ_array])
+        if save_data
+            isdir(chkptdir) || mkpath(chkptdir)
+            println("iteration: ",  iteration, "  ", iteration % checkpoint_every)
+            if iteration % checkpoint_every == 0 
+                println("CREATE CHECKPOINT AT ITERATION = ", iteration, " TIME = ", t)
+                checkpoint_filename = joinpath(chkptdir, "checkpoint.chkpt.it" * lpad(iteration, 6, "0") * ".h5")
+                checkpoint_simulation_hdf5(checkpoint_filename, s, τ, N, B, L, N_sites, Δx, Δm², p, x, Δp, gates, theta_nu, ψ, shape_name, energy_sign, cutoff, maxdim, t1, t2, ttotal, t, iteration)
+            end
 
-        fname8 = joinpath(datadir, "Im_Ω.dat")
-        writedlm(fname8, [Im_Ω])
+            iteration = iteration + 1
+        end
     end
-    
-    return Sz_array, Sy_array, Sx_array, prob_surv_array, x_values, pₓ_values, ρₑₑ_array, ρ_μμ_array, ρₑμ_array, Im_Ω 
+    t_array = t_initial:τ:ttotal
+
+    if save_data 
+        store_data(do_recover, datadir, iteration, checkpoint_every, t_array, Sz_array, Sy_array, Sx_array, prob_surv_array,x_values, pₓ_values, ρₑₑ_array,ρ_μμ_array, ρₑμ_array)
```

Comment:
I think this should be done inside of the time loop as well, via appending to files. Otherwise, if a simulation terminates, you only have the checkpoint data and not the data that you store here.